File: packages/gatekeeper/src/progress.ts
// Aggregate progress only: no per-DID identifiers or additional database reads.
export default class ProgressLogger {
    private readonly started = performance.now();
    private lastLog = this.started;

    constructor(private readonly phase: string, private readonly total: number) {
        this.log(0, this.started);
    }

    update(completed: number): void {
        const now = performance.now();
        if (completed === this.total || now - this.lastLog >= 5000) {
            this.log(completed, now);
        }
    }

    private log(completed: number, now: number): void {
        console.log(`Gatekeeper ${this.phase}: ${completed}/${this.total} DIDs (${((now - this.started) / 1000).toFixed(1)}s)`);
        this.lastLog = now;
    }
}

File: rust/services/gatekeeper/src/progress.rs
use std::time::{Duration, Instant};

/// Aggregate progress without per-DID identifiers or additional database reads.
pub(crate) struct ProgressLogger {
    phase: &'static str,
    total: usize,
    started: Instant,
    last_log: Instant,
}

impl ProgressLogger {
    pub(crate) fn new(phase: &'static str, total: usize) -> Self {
        let now = Instant::now();
        let mut progress = Self {
            phase,
            total,
            started: now,
            last_log: now,
        };
        progress.log(0, now);
        progress
    }

    pub(crate) fn update(&mut self, completed: usize) {
        let now = Instant::now();
        if completed == self.total || now.duration_since(self.last_log) >= Duration::from_secs(5) {
            self.log(completed, now);
        }
    }

    fn log(&mut self, completed: usize, now: Instant) {
        tracing::info!(
            "Gatekeeper {}: {}/{} DIDs ({:.1}s)",
            self.phase,
            completed,
            self.total,
            now.duration_since(self.started).as_secs_f64()
        );
        self.last_log = now;
    }
}
